File: typescript/src/counterparty.ts
/**
 * Counterparty acknowledgment binding helpers for the IETF Compliance Receipts profile. The binding
 * travels on the signed payload of B's `protectmcp:acknowledgment` receipt and gates verification.
 */

import { createHash } from "node:crypto";

import { canonicalJson } from "./jcs.js";

/**
 * Byte binding over payload and the exact signature object; transport_label is an operational hint.
 */
export interface CounterpartyBinding {
  envelope_hash: string;
  receipt_ref: string;
  expect_ack_from?: string;
  transport_label?: string;
  scope?: string;
}

/**
 * Per-axis outcome of the SDK-side counterparty-binding check; `label` is matches | mismatch |
 * unresolved | kid_mismatch, mirroring the cloud's `counterparty_binding_verified` axis.
 */
export interface CounterpartyBindingVerification {
  valid: boolean | null;
  envelopeHashMatches: boolean | null;
  kidMatches: boolean | null;
  label: "matches" | "mismatch" | "unresolved" | "kid_mismatch" | "malformed" | "legacy_scope" | "unrecognised_scope" | null;
}

export const BINDING_SCOPE = "envelope_minus_anchors";

function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/** Hash exactly payload and signature, preserving the encoded signature spelling. */
export function computeEnvelopeHash(
  envelope: unknown,
): string {
  if (!isObject(envelope) || !isObject(envelope.payload) || !isObject(envelope.signature)) {
    throw new Error("counterparty_origin_unavailable: signing envelope required");
  }
  const signature = envelope.signature;
  if (["alg", "kid", "sig"].some(key => typeof signature[key] !== "string" || !signature[key])) {
    throw new Error("counterparty_origin_unavailable: signature object is incomplete");
  }
  const projection = { payload: envelope.payload, signature };
  const stack: [unknown, number][] = [[projection, 0]];
  while (stack.length) {
    const [node, depth] = stack.pop()!;
    if (depth > 200) throw new Error("counterparty_origin_unavailable: nesting exceeds 200 levels");
    const children = isObject(node) ? Object.values(node) : Array.isArray(node) ? node : [];
    for (const child of children) stack.push([child, depth + 1]);
  }
  return createHash("sha256").update(canonicalJson(projection)).digest("base64");
}

/** Options for {@link computeCounterpartyBinding}. */
export interface ComputeCounterpartyBindingOptions {
  /**
   * Pass the originating signature_id for hosted admission; action_id is an offline fallback.
   */
  receiptRef?: string;
  /** Optional declared acknowledger identifier; the verifier cross-checks
   * the acknowledging receipt's `signature.kid` against this value. */
  expectAckFrom?: string;
  /** Optional operational transport hint (`mcp` | `bus` | `http`). */
  transportLabel?: string;
}

/**
 * Build from the peer's original signing envelope, retaining the signature object as received.
 */
export function computeCounterpartyBinding(
  originatingEnvelope: Record<string, unknown>,
  options: ComputeCounterpartyBindingOptions = {},
): CounterpartyBinding {
  let receiptRef = options.receiptRef;
  if (receiptRef === undefined) {
    const payload = originatingEnvelope.payload;
    if (payload && typeof payload === "object") {
      const p = payload as Record<string, unknown>;
      const candidate = p.action_id ?? p.signature_id;
      if (typeof candidate === "string") {
        receiptRef = candidate;
      }
    }
  }
  if (receiptRef === undefined) {
    throw new Error(
      "receipt_ref is required: originating envelope payload has no action_id",
    );
  }
  const binding: CounterpartyBinding = {
    envelope_hash: computeEnvelopeHash(originatingEnvelope),
    receipt_ref: receiptRef,
    scope: BINDING_SCOPE,
  };
  if (options.expectAckFrom !== undefined) {
    binding.expect_ack_from = options.expectAckFrom;
  }
  if (options.transportLabel !== undefined) {
    binding.transport_label = options.transportLabel;
  }
  return binding;
}

/**
 * Re-derive and compare the binding's `envelope_hash`, returning a label that separates a tampered
 * handoff from a missing originator. `kidMatches` is null when `expect_ack_from` was not declared.
 */
export function verifyCounterpartyBinding(
  acknowledgmentEnvelope: Record<string, unknown>,
  originatingEnvelope: unknown = null,
): CounterpartyBindingVerification {
  const result = (valid: boolean | null, label: CounterpartyBindingVerification["label"]): CounterpartyBindingVerification =>
    ({ valid, envelopeHashMatches: valid, kidMatches: null, label });
  const payload = acknowledgmentEnvelope?.payload;
  if (!isObject(payload)) {
    return result(false, "malformed");
  }
  if (!Object.hasOwn(payload, "counterparty_binding")) {
    return result(null, null);
  }
  const b = payload.counterparty_binding;
  if (!isObject(b)) return result(false, "malformed");
  if (!Object.hasOwn(b, "scope")) return result(null, "legacy_scope");
  if (b.scope !== BINDING_SCOPE) return result(null, "unrecognised_scope");
  const expectedHash = bindingDigest(b);
  if (expectedHash === null) return result(false, "malformed");
  let actualHash: Buffer;
  try {
    actualHash = Buffer.from(computeEnvelopeHash(originatingEnvelope), "base64");
  } catch {
    return result(null, "unresolved");
  }
  const envelopeHashMatches = expectedHash.equals(actualHash);

  const expectAckFrom = typeof b.expect_ack_from === "string" ? b.expect_ack_from : undefined;
  let kidMatches: boolean | null = null;
  if (expectAckFrom !== undefined) {
    const sig = acknowledgmentEnvelope.signature;
    const ackKid =
      sig && typeof sig === "object"
        ? ((sig as Record<string, unknown>).kid as unknown)
        : undefined;
    kidMatches = ackKid === expectAckFrom;
  }

  if (!envelopeHashMatches) {
    return { valid: false, envelopeHashMatches: false, kidMatches, label: "mismatch" };
  }
  if (kidMatches === false) {
    return { valid: false, envelopeHashMatches: true, kidMatches: false, label: "kid_mismatch" };
  }
  return { valid: true, envelopeHashMatches: true, kidMatches, label: "matches" };
}

function bindingDigest(binding: Record<string, unknown>): Buffer | null {
  if (typeof binding.receipt_ref !== "string" || !binding.receipt_ref) return null;
  if (["expect_ack_from", "transport_label"].some(key => binding[key] != null && typeof binding[key] !== "string")) return null;
  const value = binding.envelope_hash;
  if (typeof value !== "string" || !/^[A-Za-z0-9+/_-]+={0,2}$/.test(value) || (value.includes("=") && value.length % 4 !== 0)) return null;
  const decoded = Buffer.from(value, "base64");
  return decoded.length === 32 ? decoded : null;
}

File: python/src/asqav/counterparty.py
"""Byte bindings over the originating payload and exact signature object."""

from __future__ import annotations

import base64
import hashlib
from dataclasses import dataclass, field
from typing import Any

from ._jcs import canonical_json

__all__ = [
    "ACKNOWLEDGMENT_RECEIPT_TYPE",
    "CounterpartyBinding",
    "compute_counterparty_binding",
    "verify_counterparty_binding",
]

#: Receipt type emitted by an acknowledging agent (B) over an originating receipt.
ACKNOWLEDGMENT_RECEIPT_TYPE: str = "protectmcp:acknowledgment"
BINDING_SCOPE = "envelope_minus_anchors"


@dataclass
class CounterpartyBinding:
    """Acknowledger's cross-agent byte-equality binding to an originating receipt.

    ``transport_label`` is operational only and MUST NOT serve as a basis for
    trust derivation.
    """

    envelope_hash: str = field(
        metadata={"description": "base64 SHA-256 of originating envelope JCS bytes."}
    )
    receipt_ref: str = field(
        metadata={"description": "Opaque resolvable id for the originating receipt."}
    )
    expect_ack_from: str | None = field(
        default=None,
        metadata={"description": "Expected acknowledger kid/issuer_id."},
    )
    transport_label: str | None = field(
        default=None,
        metadata={"description": "Operational transport hint (mcp|bus|http)."},
    )
    scope: str | None = field(
        default=None,
        metadata={"description": "Digest projection; populated by the computation helper."},
    )

    def to_wire(self) -> dict[str, Any]:
        """Serialize the supplied members without assigning an unchecked digest a scope."""
        out: dict[str, Any] = {
            "envelope_hash": self.envelope_hash,
            "receipt_ref": self.receipt_ref,
        }
        if self.expect_ack_from is not None:
            out["expect_ack_from"] = self.expect_ack_from
        if self.transport_label is not None:
            out["transport_label"] = self.transport_label
        if self.scope is not None:
            out["scope"] = self.scope
        return out


def compute_envelope_hash(envelope: dict[str, Any]) -> str:
    """Hash exactly payload and signature, retaining the signature's encoded spelling."""
    if not isinstance(envelope, dict) or not isinstance(envelope.get("payload"), dict):
        raise ValueError("counterparty_origin_unavailable: payload must be an object")
    signature = envelope.get("signature")
    if not isinstance(signature, dict) or any(
        not isinstance(signature.get(key), str) or not signature[key]
        for key in ("alg", "kid", "sig")
    ):
        raise ValueError("counterparty_origin_unavailable: signature object is incomplete")
    projection = {"payload": envelope["payload"], "signature": signature}
    stack = [(projection, 0)]
    while stack:
        node, depth = stack.pop()
        if depth > 200:
            raise ValueError("counterparty_origin_unavailable: nesting exceeds 200 levels")
        children = (
            node.values() if isinstance(node, dict) else node if isinstance(node, list) else ()
        )
        stack.extend((child, depth + 1) for child in children)
    try:
        encoded = canonical_json(projection)
    except (TypeError, ValueError, RecursionError, UnicodeError) as exc:
        raise ValueError("counterparty_origin_unavailable: canonical bytes unavailable") from exc
    return base64.b64encode(hashlib.sha256(encoded).digest()).decode()


def compute_counterparty_binding(
    originating_envelope: dict[str, Any],
    *,
    receipt_ref: str | None = None,
    expect_ack_from: str | None = None,
    transport_label: str | None = None,
) -> CounterpartyBinding:
    """Build a :class:`CounterpartyBinding` for the originating envelope.

    Pass the peer's original signing envelope. Hosted admission needs an explicit
    originating signature_id as receipt_ref; the action_id fallback is an offline locator.
    """
    if receipt_ref is None:
        payload = (
            originating_envelope.get("payload") if isinstance(originating_envelope, dict) else None
        )
        if isinstance(payload, dict):
            ref = payload.get("action_id") or payload.get("signature_id")
            if isinstance(ref, str):
                receipt_ref = ref
    if receipt_ref is None:
        raise ValueError(
            "receipt_ref is required: originating_envelope.payload has no action_id"
        )
    return CounterpartyBinding(
        envelope_hash=compute_envelope_hash(originating_envelope),
        receipt_ref=receipt_ref,
        expect_ack_from=expect_ack_from,
        transport_label=transport_label,
        scope=BINDING_SCOPE,
    )


@dataclass
class CounterpartyBindingVerification:
    """Three-state byte-binding outcome; an unknown applicable binding cannot pass."""

    valid: bool | None
    envelope_hash_matches: bool | None
    kid_matches: bool | None
    label: str | None


def _binding_digest(binding: dict[str, Any]) -> bytes | None:
    """Read a well-formed binding digest under either supported base64 alphabet."""
    if not isinstance(binding.get("receipt_ref"), str) or not binding["receipt_ref"]:
        return None
    if any(binding.get(key) is not None and not isinstance(binding[key], str)
           for key in ("expect_ack_from", "transport_label")):
        return None
    value = binding.get("envelope_hash")
    if not isinstance(value, str) or not value:
        return None
    try:
        normalized = value.replace("-", "+").replace("_", "/")
        decoded = base64.b64decode(normalized + "=" * (-len(normalized) % 4), validate=True)
    except (ValueError, TypeError):
        return None
    return decoded if len(decoded) == 32 else None


def verify_counterparty_binding(
    acknowledgment_envelope: dict[str, Any],
    originating_envelope: dict[str, Any] | None = None,
) -> CounterpartyBindingVerification:
    """Dispatch scope before resolving bytes; uncertainty is separate from a mismatch."""
    payload = (
        acknowledgment_envelope.get("payload")
        if isinstance(acknowledgment_envelope, dict) else None
    )
    if not isinstance(payload, dict):
        return CounterpartyBindingVerification(False, False, None, "malformed")
    if "counterparty_binding" not in payload:
        return CounterpartyBindingVerification(None, None, None, None)
    binding = payload.get("counterparty_binding")
    if not isinstance(binding, dict):
        return CounterpartyBindingVerification(False, False, None, "malformed")
    if "scope" not in binding:
        return CounterpartyBindingVerification(None, None, None, "legacy_scope")
    if binding["scope"] != BINDING_SCOPE:
        return CounterpartyBindingVerification(None, None, None, "unrecognised_scope")
    expected_hash = _binding_digest(binding)
    if expected_hash is None:
        return CounterpartyBindingVerification(False, False, None, "malformed")
    try:
        actual_hash = base64.b64decode(compute_envelope_hash(originating_envelope))
    except (TypeError, ValueError, RecursionError):
        return CounterpartyBindingVerification(None, None, None, "unresolved")
    envelope_hash_matches = expected_hash == actual_hash
    expect_ack_from = binding.get("expect_ack_from")
    kid_matches: bool | None = None
    if isinstance(expect_ack_from, str):
        sig = acknowledgment_envelope.get("signature")
        ack_kid = sig.get("kid") if isinstance(sig, dict) else None
        kid_matches = ack_kid == expect_ack_from
    if not envelope_hash_matches:
        return CounterpartyBindingVerification(False, False, kid_matches, "mismatch")
    if kid_matches is False:
        return CounterpartyBindingVerification(False, True, False, "kid_mismatch")
    return CounterpartyBindingVerification(True, True, kid_matches, "matches")
